File: PoseKit/Tests/PoseKitTests/ScorerTests.swift
import XCTest
@testable import PoseKit

final class ScorerTests: XCTestCase {
    let standing = Fixtures.standing

    func testPerfectMatch() {
        let s = PoseScorer.score(reference: standing, live: standing)!
        XCTAssertEqual(s.overall, 1.0, accuracy: 1e-4)
        XCTAssertNil(s.hint)
    }

    func testWeighting() {
        let s = PoseScorer.score(reference: standing, live: standing)!
        XCTAssertEqual(s.overall, 0.7 * s.procrustes + 0.3 * s.limbMean, accuracy: 1e-5)
    }

    func testBadLimbProducesHint() {
        var p = standing.points
        let elbow = p[.leftElbow]!, wrist = p[.leftWrist]!
        p[.leftWrist] = elbow - (wrist - elbow)
        let s = PoseScorer.score(reference: standing, live: PoseVector(points: p))!
        XCTAssertEqual(s.worstBone, .leftForearm)
        XCTAssertEqual(s.hint, "adjust your left arm")
    }

    func testInsufficientJointsReturnsNil() {
        let tiny = PoseVector(points: [.nose: [0.5, 0.1]])
        XCTAssertNil(PoseScorer.score(reference: standing, live: tiny))
    }

    func testScoreIsTransformInvariant() {
        let t = Fixtures.transformed(standing, scale: 1.8, rotation: 0.3, translation: [2, 1])
        let s = PoseScorer.score(reference: standing, live: t)!
        // Procrustes invariant; limb cosine changes under rotation — overall still high.
        XCTAssertGreaterThan(s.overall, 0.9)
    }
}

File: android/posekit/src/test/kotlin/com/oerol/posekit/ScorerTests.kt
package com.oerol.posekit

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ScorerTests {
    private val standing = Fixtures.standing

    @Test
    fun perfectMatch() {
        val s = PoseScorer.score(standing, standing)!!
        assertEquals(1.0f, s.overall, 1e-4f)
        assertNull(s.hint)
    }

    @Test
    fun weighting() {
        val s = PoseScorer.score(standing, standing)!!
        assertEquals(0.7f * s.procrustes + 0.3f * s.limbMean, s.overall, 1e-5f)
    }

    @Test
    fun badLimbProducesHint() {
        val p = standing.points.toMutableMap()
        val elbow = p.getValue(Joint.leftElbow)
        val wrist = p.getValue(Joint.leftWrist)
        p[Joint.leftWrist] = elbow - (wrist - elbow)
        val s = PoseScorer.score(standing, PoseVector(p))!!
        assertEquals(Bone.LEFT_FOREARM, s.worstBone)
        assertEquals("adjust your left arm", s.hint)
    }

    @Test
    fun insufficientJointsReturnsNull() {
        val tiny = PoseVector(mapOf(Joint.nose to Vec2(0.5f, 0.1f)))
        assertNull(PoseScorer.score(standing, tiny))
    }

    @Test
    fun scoreIsTransformInvariant() {
        val t = Fixtures.transformed(standing, scale = 1.8f, rotation = 0.3f, translation = Vec2(2f, 1f))
        val s = PoseScorer.score(standing, t)!!
        // Procrustes invariant; limb cosine changes under rotation — overall still high.
        assertTrue(s.overall > 0.9f)
    }
}
